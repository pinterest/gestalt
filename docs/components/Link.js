// @flow strict
import { type Node } from 'react';
import { Link as GestaltLink } from 'gestalt';
import { useRouter } from 'next/router';

type Props = {|
  accessibilityLabel?: $ElementType<React$ElementConfig<typeof GestaltLink>, 'accessibilityLabel'>,
  children?: $ElementType<React$ElementConfig<typeof GestaltLink>, 'children'>,
  href: $ElementType<React$ElementConfig<typeof GestaltLink>, 'href'>,
  onClick?: () => void,
  target?: $ElementType<React$ElementConfig<typeof GestaltLink>, 'target'>,
|};

const isModifiedEvent = (event) =>
  !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);

export default function Link({ accessibilityLabel, children, href, target, onClick }: Props): Node {
  const router = useRouter();

  const handleClick = ({ event }) => {
    onClick?.();

    if (
      !event.defaultPrevented && // onClick prevented default
      event.button === 0 && // ignore everything but left clicks
      !target && // let browser handle "target=_blank" etc.
      !isModifiedEvent(event) // ignore clicks with modifier keys
    ) {
      event.preventDefault();

      router.push(href);
    }
  };

  return (
    <GestaltLink
      accessibilityLabel={accessibilityLabel}
      href={href}
      onClick={handleClick}
      target={target}
    >
      {children}
    </GestaltLink>
  );
}
