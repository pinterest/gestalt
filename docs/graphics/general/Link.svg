<svg xmlns="http://www.w3.org/2000/svg" width="312" height="180" fill="none"><path fill="#005C62" d="M146.544 93.12h-6.594V81.046h-2.185v13.962h8.779v-1.886Zm3.609-10.498a1.32 1.32 0 0 0 1.326-1.306c0-.726-.595-1.316-1.326-1.316-.732 0-1.327.59-1.327 1.316a1.32 1.32 0 0 0 1.327 1.306Zm-1.044 12.385h2.087V84.5h-2.087v10.508Zm5.169 0h2.098v-6.163c0-1.694.985-2.738 2.535-2.738 1.551 0 2.293.85 2.293 2.593v6.308h2.097v-6.802c0-2.506-1.307-3.909-3.677-3.909-1.6 0-2.654.706-3.17 1.867h-.157V84.5h-2.019v10.508Zm14.338-6.077h-.165v-8.553h-2.097v14.63h2.097v-3.822l.926-.88 3.648 4.702h2.575L170.938 89l4.369-4.5h-2.477l-4.214 4.432ZM136 98.597h40V100h-40v-1.403Z"/></svg>