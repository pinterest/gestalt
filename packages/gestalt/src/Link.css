.link {
  composes: noUnderline from "./Typography.css";
  color: inherit; /* Links should have the correct underline color */
  outline: none;
}

.link:hover,
.link:focus {
  text-decoration: underline;
}

.accessibleFocusStyle {
  composes: accessibilityOutline from "./Focus.css";
}

.block {
  composes: block from "./Layout.css";
}
