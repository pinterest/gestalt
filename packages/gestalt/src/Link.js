// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './Link.css';

type Props = {|
  children?: React.Node,
  href: string,
  inline?: boolean,
  onClick?: ({ event: SyntheticMouseEvent<> }) => void,
  target?: null | 'self' | 'blank',
|};

type State = {|
  enableFocusStyles: boolean,
|};

const TAB_KEY_CODE = 9;

export default class Link extends React.Component<Props, State> {
  static propTypes = {
    children: PropTypes.node.isRequired,
    href: PropTypes.string.isRequired,
    inline: PropTypes.bool,
    onClick: PropTypes.func,
    target: PropTypes.oneOf([null, 'self', 'blank']),
  };

  state: State = {
    enableFocusStyles: true,
  };

  handleClick = (event: SyntheticMouseEvent<>) => {
    const { href, onClick } = this.props;
    if (onClick && href) {
      onClick({ event });
    }
  };

  handleMouseDown = () => {
    const { href, target } = this.props;
    if (target === 'blank' && href) {
      this.setState({ enableFocusStyles: false });
    }
  };

  handleKeyUp = (event: SyntheticKeyboardEvent<>) => {
    const { href, target } = this.props;
    if (target === 'blank' && event.keyCode === TAB_KEY_CODE && href) {
      this.setState({ enableFocusStyles: true });
    }
  };

  render() {
    const { children, inline = false, target = null, href } = this.props;
    const rel = target === 'blank' ? 'noopener noreferrer' : null;
    const linkTarget = target ? `_${target}` : null;

    return (
      <a
        className={cx(
          styles.link,
          this.state.enableFocusStyles ? styles.accessibleFocusStyle : '',
          inline ? '' : styles.block
        )}
        href={href}
        onMouseDown={this.handleMouseDown}
        onKeyUp={this.handleKeyUp}
        onClick={this.handleClick}
        rel={rel}
        target={linkTarget}
      >
        {children}
      </a>
    );
  }
}
