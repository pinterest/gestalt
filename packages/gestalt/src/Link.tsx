import {
  ComponentProps,
  forwardRef,
  ReactElement,
  ReactNode,
  Ref,
  useImperativeHandle,
  useRef,
} from 'react';
import classnames from 'classnames';
import AccessibilityLinkActionIcon from './accessibility/AccessibilityLinkActionIcon';
import getAriaLabel from './accessibility/getAriaLabel';
import Box from './Box';
import { useDefaultLabelContext } from './contexts/DefaultLabelProvider';
import { useGlobalEventsHandlerContext } from './contexts/GlobalEventsHandlerProvider';
import focusStyles from './Focus.css';
import getRoundingClassName from './getRoundingClassName';
import layoutStyles from './Layout.css';
import touchableStyles from './TapArea.css';
import Text from './Text';
import styles from './Text.css';
import typographyStyles from './Typography.css';
import useFocusVisible from './useFocusVisible';
import useInExperiment from './useInExperiment';
import useTapFeedback, { keyPressShouldTriggerTap } from './useTapFeedback';

const externalLinkIconMap = {
  '100': 12,
  '200': 14,
  '300': 16,
  '400': 20,
  '500': 28,
  '600': 36,
  sm: 12,
  md: 14,
  lg: 16,
} as const;

type ExternalLinkIcon =
  | 'none'
  | 'default'
  | {
      size: ComponentProps<typeof Text>['size'];
    };

type Rounding = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 'circle' | 'pill';

type Props = {
  /**
   * Supply a short, descriptive label for screen-readers to replace link texts that don't provide sufficient context about the link component behavior. Texts like "Click Here", or "Read More" can be confusing when a screen reader reads them out of context. In those cases, we must pass an alternative text to replace the link text. It populates `aria-label`. Screen readers read the `accessibilityLabel` prop, if present, instead of the link text. See the [Accessibility guidelines](https://gestalt.pinterest.systems/web/link#Accessibility) for more information.
   */
  accessibilityLabel?: string;
  /**
   * Link is a wrapper around components (or children), most commonly text, so that they become hyperlinks. See the [Text and Link variant](https://gestalt.pinterest.systems/web/link#Link-and-Text) to learn more.
   */
  children?: ReactNode;
  /**
   * Available for testing purposes, if needed. Consider [better queries](https://testing-library.com/docs/queries/about/#priority) before using this prop.
   */
  dataTestId?: string;
  /**
   * Determines how Link is positioned relative to surrounding elements, such as [Text](https://gestalt.pinterest.systems/web/text). See the [inline variant](https://gestalt.pinterest.systems/web/link#Inline) to learn more.
   */
  display?: 'inline' | 'inlineBlock' | 'block';
  /**
   * When supplied, a "visit" icon is shown at the end of Link. See the [externalLinkIcon and rel variant](https://gestalt.pinterest.systems/web/link#externalLinkIcon-and-rel) to learn more.
   */
  externalLinkIcon?: ExternalLinkIcon;
  /**
   * The URL that the hyperlink points to.
   */
  href: string;
  /**
   * Unique id attribute of the anchor tag.
   */
  id?: string;
  /**
   * Callback triggered when when the element loses focus.
   */
  onBlur?: (arg1: { event: React.FocusEvent<HTMLAnchorElement> }) => void;
  /**
   * Callback fired when Link is clicked (pressed and released) with a mouse or keyboard. See [GlobalEventsHandlerProvider](https://gestalt.pinterest.systems/web/utilities/globaleventshandlerprovider#Link-handlers) to learn more about link navigation.
   */
  onClick?: (arg1: {
    event: React.MouseEvent<HTMLAnchorElement> | React.KeyboardEvent<HTMLAnchorElement>;
    dangerouslyDisableOnNavigation: () => void;
  }) => void;
  /**
   * Callback triggered when the element gains focus.
   */
  onFocus?: (arg1: { event: React.FocusEvent<HTMLAnchorElement> }) => void;
  /**
   * Ref that is forwarded to the underlying anchor element.
   */
  ref?: Ref<'a'>;
  /**
   * Establishes the relationship of the linked URL. Use `rel="nofollow"` for offsite links to inform search engines to ignore and not follow them. See the [externalLinkIcon and rel variant](https://gestalt.pinterest.systems/web/link#externalLinkIcon-and-rel) to learn more.
   */
  rel?: 'none' | 'nofollow';
  /**
   * Sets a border radius for Link. Select a rounding option that aligns with its children.
   */
  rounding?: Rounding;
  /**
   * When `compress` is supplied, Link is visually compressed on click.
   */
  tapStyle?: 'none' | 'compress';
  /**
     * Define the frame or window to open the anchor defined on href:
  - 'null' opens the anchor in the same window.
  - 'blank' opens the anchor in a new window.
  - 'self' opens an anchor in the same frame.

  See the [target variant](https://gestalt.pinterest.systems/web/link#Target) to learn more.
     */
  target?: null | 'self' | 'blank';
  /**
   * When `underline` is supplied, we override the underline style internally managed by the component. See the [underline variant](https://gestalt.pinterest.systems/web/link#Underline) to learn more.
   */
  underline?: 'auto' | 'none' | 'always' | 'hover';
};

/**
 * [Link](https://gestalt.pinterest.systems/web/link) is mainly used as navigational element and usually appear within or directly following a paragraph or sentence.
 *
 * ![Link light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Link.spec.ts-snapshots/Link-chromium-darwin.png)
 * ![Link dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Link-dark.spec.ts-snapshots/Link-dark-chromium-darwin.png)
 *
 */
const LinkWithForwardRef = forwardRef<HTMLAnchorElement, Props>(function Link(
  {
    accessibilityLabel,
    children,
    dataTestId,
    display = 'block',
    externalLinkIcon = 'none',
    href,
    id,
    onBlur,
    onClick,
    onFocus,
    rel = 'none',
    rounding = 0,
    underline = 'auto',
    tapStyle = 'none',
    target = null,
  }: Props,
  ref,
): ReactElement {
  const innerRef = useRef<null | HTMLAnchorElement>(null);

  const isInVRExperiment = useInExperiment({
    webExperimentName: 'web_gestalt_visualRefresh',
    mwebExperimentName: 'web_gestalt_visualRefresh',
  });

  // @ts-expect-error - TS2322 - Type 'HTMLAnchorElement | null' is not assignable to type 'HTMLAnchorElement'.
  useImperativeHandle(ref, () => innerRef.current);

  const {
    compressStyle,
    isTapping,
    handleBlur,
    handleMouseDown,
    handleMouseUp,
    handleTouchStart,
    handleTouchMove,
    handleTouchCancel,
    handleTouchEnd,
  } = useTapFeedback({
    height: innerRef?.current?.clientHeight,
    width: innerRef?.current?.clientWidth,
  });

  const { accessibilityNewTabLabel } = useDefaultLabelContext('Link');

  const { isFocusVisible } = useFocusVisible();

  const isInline = ['inline', 'inlineBlock'].includes(display);

  let underlineStyle = isInline ? 'always' : 'hover';

  if (underline && underline !== 'auto') {
    underlineStyle = underline;
  }

  const className = classnames(
    styles.noOutline,
    styles.inheritColor,
    getRoundingClassName(rounding),
    layoutStyles[display],
    touchableStyles.tapTransition,
    {
      [styles.hoverNoUnderline]: underlineStyle === 'always',
      [styles.hoverUnderline]: underlineStyle === 'hover',
      [styles.underline]: underlineStyle === 'always',
      [styles.noUnderline]: underlineStyle === 'hover' || underlineStyle === 'none',
      [styles.outlineFocus]: isFocusVisible,
      [focusStyles.hideOutline]: !isFocusVisible,
      [touchableStyles.tapCompress]: tapStyle === 'compress' && isTapping,
    },
  );

  const VRclassName = classnames(
    styles.noOutline,
    styles.inheritColor,
    getRoundingClassName(rounding),
    layoutStyles[display],
    touchableStyles.tapTransition,
    {
      [styles.vrInheritColor]: isInline,
      [styles.standalone]: !isInline,
      [styles.underline]: underlineStyle === 'always',
      [styles.noUnderline]: underlineStyle === 'hover' || underlineStyle === 'none',
      [styles.hoverUnderline]: underlineStyle === 'hover',
      [styles.outlineFocusVR]: isFocusVisible,
      [typographyStyles.fontWeightSemiBold]: !isInline,
      [focusStyles.hideOutline]: !isFocusVisible,
      [touchableStyles.tapCompress]: tapStyle === 'compress' && isTapping,
    },
  );

  // Consumes GlobalEventsHandlerProvider
  const { linkHandlers } = useGlobalEventsHandlerContext() ?? {
    linkHandlers: { onNavigation: undefined },
  };

  const { onNavigation } = linkHandlers ?? { onNavigation: undefined };

  const onNavigationHandler = onNavigation?.({ href, target });

  const handleKeyPress = (event: React.KeyboardEvent<HTMLAnchorElement>) => {
    // Check to see if space or enter were pressed
    if (onClick && keyPressShouldTriggerTap(event)) {
      // Prevent the default action to stop scrolling when space is pressed
      event.preventDefault();
      onClick({ event, dangerouslyDisableOnNavigation: () => {} });
    }
  };

  const ariaLabel = getAriaLabel({
    target,
    accessibilityLabel,
    accessibilityNewTabLabel,
  });

  return (
    <a
      ref={innerRef}
      aria-label={ariaLabel}
      className={isInVRExperiment ? VRclassName : className}
      data-test-id={dataTestId}
      href={href}
      id={id}
      onBlur={(event) => {
        handleBlur();
        onBlur?.({ event });
      }}
      onClick={(event) => {
        let defaultOnNavigationIsEnabled = true;
        const dangerouslyDisableOnNavigation = () => {
          defaultOnNavigationIsEnabled = false;
        };

        onClick?.({
          event,
          dangerouslyDisableOnNavigation,
        });

        if (onNavigationHandler && defaultOnNavigationIsEnabled) {
          onNavigationHandler({ event });
        }
      }}
      onFocus={(event) => {
        onFocus?.({ event });
      }}
      onKeyPress={handleKeyPress}
      onMouseDown={handleMouseDown}
      onMouseUp={handleMouseUp}
      onTouchCancel={handleTouchCancel}
      onTouchEnd={handleTouchEnd}
      onTouchMove={handleTouchMove}
      onTouchStart={handleTouchStart}
      rel={[
        ...(target === 'blank' ? ['noopener', 'noreferrer'] : []),
        ...(rel === 'nofollow' ? ['nofollow'] : []),
      ].join(' ')}
      {...(compressStyle && tapStyle === 'compress' ? { style: compressStyle } : {})}
      target={target ? `_${target}` : undefined}
    >
      {children}
      {externalLinkIcon === 'none' ? null : (
        <Box display="inlineBlock" marginStart={1}>
          <AccessibilityLinkActionIcon
            icon="visit"
            size={
              externalLinkIcon === 'default'
                ? externalLinkIconMap['300']
                : externalLinkIconMap[externalLinkIcon?.size ?? '300']
            }
          />
        </Box>
      )}
    </a>
  );
});

LinkWithForwardRef.displayName = 'Link';

export default LinkWithForwardRef;
