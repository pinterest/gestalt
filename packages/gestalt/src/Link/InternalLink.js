// @flow strict
import {
  type AbstractComponent,
  type Element,
  forwardRef,
  type Node,
  useImperativeHandle,
  useRef,
} from 'react';
import classnames from 'classnames';
import { type AriaCurrent } from '../ariaTypes.js';
import buttonStyles from '../Button.css';
import { useGlobalEventsHandlerContext } from '../contexts/GlobalEventsHandlerProvider.js';
import focusStyles from '../Focus.css';
import getRoundingClassName, { type Rounding } from '../getRoundingClassName.js';
import iconButtonStyles from '../IconButton.css';
import layoutStyles from '../Layout.css';
import linkStyles from '../Link.css';
import touchableStyles from '../TapArea.css';
import textStyles from '../Typography.css';
import useFocusVisible from '../useFocusVisible.js';
import useTapFeedback, { keyPressShouldTriggerTap } from '../useTapFeedback.js';

type Props = {
  accessibilityCurrent?: AriaCurrent,
  accessibilityLabel?: string,
  children?: Node,
  colorClass?: string,
  dataTestId?: string,
  disabled?: boolean,
  fullHeight?: boolean,
  fullWidth?: boolean,
  href: string,
  id?: string,
  mouseCursor?: 'copy' | 'grab' | 'grabbing' | 'move' | 'noDrop' | 'pointer' | 'zoomIn' | 'zoomOut',
  onClick?: ({
    event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticKeyboardEvent<HTMLAnchorElement>,
    dangerouslyDisableOnNavigation: () => void,
  }) => void,
  onBlur?: ({
    event: SyntheticFocusEvent<HTMLAnchorElement>,
  }) => void,
  onFocus?: ({
    event: SyntheticFocusEvent<HTMLAnchorElement>,
  }) => void,
  onKeyDown?: ({
    event: SyntheticKeyboardEvent<HTMLAnchorElement>,
  }) => void,
  onMouseDown?: ({
    event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticMouseEvent<HTMLDivElement>,
  }) => void,
  onMouseUp?: ({
    event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticMouseEvent<HTMLDivElement>,
  }) => void,
  onMouseEnter?: ({
    event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticMouseEvent<HTMLDivElement>,
  }) => void,
  onMouseLeave?: ({
    event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticMouseEvent<HTMLDivElement>,
  }) => void,
  rel?: 'none' | 'nofollow',
  tabIndex: -1 | 0,
  rounding?: Rounding,
  selected?: boolean,
  size?: 'sm' | 'md' | 'lg',
  tapStyle?: 'none' | 'compress',
  target?: null | 'self' | 'blank',
  wrappedComponent: 'button' | 'iconButton' | 'tapArea',
};

const InternalLinkWithForwardRef: AbstractComponent<Props, HTMLAnchorElement> = forwardRef<
  Props,
  HTMLAnchorElement,
>(function Link(
  {
    accessibilityCurrent,
    accessibilityLabel,
    children,
    colorClass,
    dataTestId,
    disabled,
    fullHeight,
    fullWidth,
    href,
    id,
    mouseCursor,
    onClick,
    onBlur,
    onFocus,
    onKeyDown,
    onMouseDown,
    onMouseUp,
    onMouseEnter,
    onMouseLeave,
    rel,
    tabIndex = 0,
    rounding,
    selected,
    size,
    tapStyle = 'compress',
    target,
    wrappedComponent,
  }: Props,
  ref,
): Element<'a'> {
  const innerRef = useRef<null | HTMLAnchorElement>(null);

  useImperativeHandle(ref, () => innerRef.current);

  const {
    compressStyle,
    handleBlur,
    handleMouseDown,
    handleMouseUp,
    handleTouchStart,
    handleTouchMove,
    handleTouchCancel,
    handleTouchEnd,
    isTapping,
  } = useTapFeedback({
    height: innerRef?.current?.clientHeight,
    width: innerRef?.current?.clientWidth,
  });

  const { isFocusVisible } = useFocusVisible();
  const isTapArea = wrappedComponent === 'tapArea';
  const isButton = wrappedComponent === 'button';
  const isIconButton = wrappedComponent === 'iconButton';

  const className = classnames(
    linkStyles.link,
    textStyles.noUnderline,
    focusStyles.hideOutline,
    touchableStyles.tapTransition,
    getRoundingClassName(isTapArea ? rounding || 0 : 'pill'),
    {
      [touchableStyles.tapCompress]: !disabled && tapStyle === 'compress' && isTapping,
      [focusStyles.accessibilityOutline]: !disabled && isFocusVisible,
    },
    isButton
      ? {
          [layoutStyles.inlineFlex]: !fullWidth,
          [layoutStyles.flex]: fullWidth,
          [layoutStyles.justifyCenter]: true,
          [layoutStyles.xsItemsCenter]: true,
          [buttonStyles.button]: true,
          [buttonStyles.disabled]: disabled,
          [buttonStyles.selected]: !disabled && selected,
          [buttonStyles.sm]: size === 'sm',
          [buttonStyles.md]: size === 'md',
          [buttonStyles.lg]: size === 'lg',
        }
      : {},
    isButton && colorClass
      ? {
          [buttonStyles[colorClass]]: !disabled && !selected,
        }
      : {},
    isTapArea
      ? {
          [layoutStyles.block]: true,
          [touchableStyles.fullHeight]: fullHeight,
          [touchableStyles.fullWidth]: fullWidth,
        }
      : {},
    isTapArea && mouseCursor
      ? {
          // $FlowFixMe[invalid-computed-prop]
          [touchableStyles[mouseCursor]]: !disabled,
        }
      : {},
    isIconButton
      ? {
          [iconButtonStyles.button]: true,
          [iconButtonStyles.disabled]: disabled,
          [iconButtonStyles.enabled]: !disabled,
        }
      : {},
  );

  // Consumes GlobalEventsHandlerProvider
  const { linkHandlers } = useGlobalEventsHandlerContext() ?? {
    linkHandlers: { onNavigation: undefined },
  };

  const { onNavigation } = linkHandlers ?? { onNavigation: undefined };

  const onNavigationHandler = onNavigation?.({ href, target });

  const handleKeyPress = (event: SyntheticKeyboardEvent<HTMLAnchorElement>) => {
    // Check to see if space or enter were pressed
    if (onClick && keyPressShouldTriggerTap(event)) {
      // Prevent the default action to stop scrolling when space is pressed
      event.preventDefault();
      onClick({ event, dangerouslyDisableOnNavigation: () => {} });
    }
  };

  return (
    <a
      aria-current={accessibilityCurrent !== 'section' ? accessibilityCurrent : undefined}
      aria-selected={accessibilityCurrent === 'section' ? accessibilityCurrent : undefined}
      aria-label={accessibilityLabel}
      className={className}
      data-test-id={dataTestId}
      href={disabled ? undefined : href}
      id={id}
      onBlur={(event) => {
        onBlur?.({ event });
        handleBlur();
      }}
      onClick={(event) => {
        let defaultOnNavigationIsEnabled = true;
        const dangerouslyDisableOnNavigation = () => {
          defaultOnNavigationIsEnabled = false;
        };

        onClick?.({
          event,
          dangerouslyDisableOnNavigation,
        });
        if (onNavigationHandler && defaultOnNavigationIsEnabled) {
          onNavigationHandler({ event });
        }
      }}
      onFocus={(event) => {
        onFocus?.({ event });
      }}
      onKeyDown={(event) => {
        onKeyDown?.({ event });
      }}
      onMouseEnter={(event) => {
        onMouseEnter?.({ event });
      }}
      onMouseLeave={(event) => {
        onMouseLeave?.({ event });
      }}
      onMouseDown={(event) => {
        onMouseDown?.({ event });
        handleMouseDown();
      }}
      onMouseUp={(event) => {
        onMouseUp?.({ event });
        handleMouseUp();
      }}
      onKeyPress={handleKeyPress}
      onTouchStart={handleTouchStart}
      onTouchMove={handleTouchMove}
      onTouchCancel={handleTouchCancel}
      onTouchEnd={handleTouchEnd}
      ref={innerRef}
      rel={[
        ...(target === 'blank' ? ['noopener', 'noreferrer'] : []),
        ...(rel === 'nofollow' ? ['nofollow'] : []),
      ].join(' ')}
      tabIndex={disabled ? null : tabIndex}
      {...(tapStyle === 'compress' && compressStyle && !disabled ? { style: compressStyle } : {})}
      target={target ? `_${target}` : null}
    >
      {children}
    </a>
  );
});

InternalLinkWithForwardRef.displayName = 'InternalLink';

export default InternalLinkWithForwardRef;
