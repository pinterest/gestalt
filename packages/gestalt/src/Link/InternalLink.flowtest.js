// @flow strict
import InternalLink from './InternalLink';

const Valid = (
  <InternalLink href="https://example.com" tabIndex={0} wrappedComponent="button">
    content
  </InternalLink>
);

// $FlowExpectedError[prop-missing]
const MissingProp = <InternalLink />;

// $FlowExpectedError[prop-missing]
const NonExistingProp = <InternalLink nonexisting={33} />;
