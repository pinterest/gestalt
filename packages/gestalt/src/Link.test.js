// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import Link from './Link.js';

const snapshot = component => {
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
};

it('default', () => snapshot(<Link href="https://example.com">Link</Link>));
it('regular', () => snapshot(<Link href="https://example.com">Link</Link>));
it('inline', () =>
  snapshot(
    <Link href="https://example.com" inline>
      Link
    </Link>
  ));

it('target null', () =>
  snapshot(
    <Link href="https://example.com" target={null}>
      Link
    </Link>
  ));

it('target self', () =>
  snapshot(
    <Link href="https://example.com" target="self">
      Link
    </Link>
  ));

it('target blank', () =>
  snapshot(
    <Link href="https://example.com" target="blank">
      Link
    </Link>
  ));

it('with onClick', () =>
  snapshot(
    <Link href="https://example.com" onClick={() => {}}>
      Link
    </Link>
  ));
