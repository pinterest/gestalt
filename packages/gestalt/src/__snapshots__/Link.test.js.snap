// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
<a
  className="link tapTransition touchable block rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel=""
  target={null}
>
  Link
</a>
`;

exports[`inline 1`] = `
<a
  className="link tapTransition touchable inlineBlock rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel=""
  target={null}
>
  Link
</a>
`;

exports[`regular 1`] = `
<a
  className="link tapTransition touchable block rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel=""
  target={null}
>
  Link
</a>
`;

exports[`target blank 1`] = `
<a
  className="link tapTransition touchable block rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel="noopener noreferrer"
  target="_blank"
>
  Link
</a>
`;

exports[`target null 1`] = `
<a
  className="link tapTransition touchable block rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel=""
  target={null}
>
  Link
</a>
`;

exports[`target self 1`] = `
<a
  className="link tapTransition touchable block rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel=""
  target="_self"
>
  Link
</a>
`;

exports[`with accessibilitySelected and role 1`] = `
<a
  aria-selected={true}
  className="link tapTransition touchable block rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel=""
  role="tab"
  target={null}
>
  Link
</a>
`;

exports[`with custom rounding, hoverStyle, and tapStyle 1`] = `
<a
  className="link tapTransition touchable block pill"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel=""
  target={null}
>
  Link
</a>
`;

exports[`with nofollow 1`] = `
<a
  className="link tapTransition touchable block rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel="nofollow"
  target={null}
>
  Link
</a>
`;

exports[`with onTap 1`] = `
<a
  className="link tapTransition touchable block rounding0 hoverUnderline"
  href="https://example.com"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  rel=""
  target={null}
>
  Link
</a>
`;
